src/pj_fwd.c

```diff
@@ -1,58 +1,174 @@
-            lp.phi = lp.phi < 0. ? -M_HALFPI : M_HALFPI;
-        else if (P->geoc)   /* Maybe redundant and never used. */
-            lp.phi = atan(P->rone_es * tan(lp.phi));
-        lp.lam -= P->lam0;    /* compute del lp.lam */
-        if (!P->over)
-            lp.lam = adjlon(lp.lam); /* adjust del longitude */
+        if (lp.phi > M_HALFPI)
+            lp.phi = M_HALFPI;
+        if (lp.phi < -M_HALFPI)
+            lp.phi = -M_HALFPI;
+
+        /* If input latitude is geocentrical, convert to geographical */
+        if (P->geoc)
+            coo = proj_geoc_lat (P, PJ_INV, coo);
+
+        /* Distance from central meridian, taking system zero meridian into account */
+        lp.lam = (lp.lam - P->from_greenwich) - P->lam0;
+
+        /* Ensure longitude is in the -pi:pi range */
+        if (0==P->over)
+            lp.lam = adjlon(lp.lam);
+
+        coo.lp = lp;
     }
 
-    /* Do the transformation */
-    xy = (*P->fwd)(lp, P);
-    if ( proj_errno (P) )
-        return err;
+    return coo;
+}
+
+
+PJ_COORD pj_fwd_finalize (PJ *P, PJ_COORD coo) {
 
     /* Classic proj.4 functions return plane coordinates in units of the semimajor axis */
     if (P->right==PJ_IO_UNITS_CLASSIC) {
-        xy.x *= P->a;
-        xy.y *= P->a;
+        coo.xy.x *= P->a;
+        coo.xy.y *= P->a;
     }
 
     /* Handle false eastings/northings and non-metric linear units */
-    xy.x = P->fr_meter * (xy.x + P->x0);
-    xy.y = P->fr_meter * (xy.y + P->y0);
-    /* z is not scaled since this is handled by vto_meter outside */
+    coo.xyz.x = P->fr_meter  * (coo.xyz.x + P->x0);
+    coo.xyz.y = P->fr_meter  * (coo.xyz.y + P->y0);
+    coo.xyz.z = P->vfr_meter * (coo.xyz.z + P->z0);
+
+    return coo;
+}
+
+
+
+XY pj_fwd(LP lp, PJ *P) {
+    PJ_COORD coo = {{0,0,0,0}};
+    coo.lp = lp;
+
+    if (!P->skip_fwd_prepare)
+        coo = pj_fwd_prepare (P, coo);
+    if (HUGE_VAL==coo.v[0])
+        return proj_coord_error ().xy;
+
+    /* Do the transformation, using the lowest dimensional transformer available */
+    if (P->fwd)
+        coo.xy = P->fwd(coo.lp, P);
+    else if (P->fwd3d)
+        coo.xyz = P->fwd3d (coo.lpz, P);
+    else if (P->fwd4d)
+        coo = P->fwd4d (coo, P);
+    else {
+        proj_errno_set (P, EINVAL);
+        return proj_coord_error ().xy;
+    }
+    if (HUGE_VAL==coo.v[0])
+        return proj_coord_error ().xy;
+
+    if (!P->skip_fwd_finalize)
+        coo = pj_fwd_finalize (P, coo);
+    return coo.xy;
+}
+
+
+
+XYZ pj_fwd3d(LPZ lpz, PJ *P) {
+    PJ_COORD coo = {{0,0,0,0}};
+    coo.lpz = lpz;
+
+    if (!P->skip_fwd_prepare)
+        coo = pj_fwd_prepare (P, coo);
+    if (HUGE_VAL==coo.v[0])
+        return proj_coord_error ().xyz;
+
+    /* Do the transformation, using the lowest dimensional transformer feasible */
+    if (P->fwd3d)
+        coo.xyz = P->fwd3d(coo.lpz, P);
+    else if (P->fwd4d)
+        coo = P->fwd4d (coo, P);
+    else {
+        proj_errno_set (P, EINVAL);
```

Comment:
Long story, which I will move out to the discussion level of the thread, since it does not really fit into a review comment.

Reply:
Changed my mind on the previous comment.

For now, I'll let it fall back to 2D eventually, but I will draft up an issue describing why the 3D low level interfaces should be abandoned altogether:

Their actual functionality can be handled by 4D, and except for our geodetic extensions (which could all be slightly simplified if we reduced them to 4D), the 3D entries were introduced in connection with `PJ_sch.c`, which is also the only other operation supporting it (and it could easily be changed to work in 4D). All new functionality should recognize that the world is 4D.

But that's another banana for another PR

Reply:
This seem like a reasonable approach to me.